torch/csrc/jit/codegen/cuda/executor.cpp

```diff
@@ -283,42 +283,106 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
 std::vector<at::Tensor> FusionExecutor::runFusion(
     const at::ArrayRef<IValue>& inputs,
     const std::vector<at::Tensor>& outputs,
-    const LaunchParams& launch_constraints) {
+    const LaunchParams& launch_constraints,
+    const c10::optional<size_t>& opt_code) {
   TORCH_INTERNAL_ASSERT(
       fusion_id_ > 0, "Cannot run fusion, it was not compiled.");
+  TORCH_INTERNAL_ASSERT(
+      !opt_code.has_value() || outputs.empty(),
+      "short cut input cache is not compatible with pre-allocated output");
 
-  FusionGuard fg(&fusion_);
-
-  executor_utils::validateKernelInputs(&fusion_, inputs, options_.device);
+  ExecutorEntry* executor_entry = nullptr;
+  if (opt_code.has_value()) {
+    executor_entry = &executor_entry_lookup_[*opt_code];
+  }
 
+  FusionGuard fg(&fusion_);
   c10::DeviceGuard dg(options_.device);
   auto stream = at::cuda::getCurrentCUDAStream();
 
-  EvaluationContext evaluation_context =
-      executor_utils::bindInputs(inputs, &fusion_, &lowered_);
+  LaunchParams launch_params;
+  std::vector<at::Tensor> alloced_outputs = outputs;
+  std::vector<at::Tensor> empty_buffers;
+  std::vector<at::Tensor> zero_buffers;
+  uint64_t rand_offset = 0;
+
+  if (executor_entry && executor_entry->init) {
+    // take the short-cut for launch if we see a recorded input set again;
+    launch_params = executor_entry->launch_params;
+    for (size_t i = 0; i < executor_entry->output_sizes.size(); i++) {
+      auto tensor_options = at::TensorOptions()
+                                .dtype(executor_entry->output_types[i])
+                                .device(options_.device);
+      alloced_outputs.push_back(
+          at::empty(executor_entry->output_sizes[i], tensor_options));
+    }
+    for (size_t i = 0; i < executor_entry->empty_buffer_sizes.size(); i++) {
+      auto tensor_options = at::TensorOptions()
+                                .dtype(executor_entry->empty_buffer_types[i])
+                                .device(options_.device);
+      empty_buffers.push_back(
+          at::empty(executor_entry->empty_buffer_sizes[i], tensor_options));
+    }
+    for (size_t i = 0; i < executor_entry->zero_buffer_sizes.size(); i++) {
+      auto tensor_options = at::TensorOptions()
+                                .dtype(executor_entry->zero_buffer_types[i])
+                                .device(options_.device);
+      zero_buffers.push_back(
+          at::zeros(executor_entry->zero_buffer_sizes[i], tensor_options));
```

Comment:
at::empty & at::zeros go through the cuda caching allocator in PyTorch. So they maintain their own memory pools.

> for each buffer, it will launch a kernel to zero-initialize the device memory

I don't know how else that we can get 0-initialized memory other than a set memory, iiuc, which is a kernel by itself.

Reply:
We can combine all the allocations into a single cudaMalloc() + cudaMemset().

Reply:
I think this is going to be a longer discussion. I'll open an issue to track it.

Reply:
https://github.com/csarofeen/pytorch/issues/332